source/transfers/test/TransferHandlerRegistry.js

```diff
@@ -0,0 +1,1008 @@
+        ])
+      )
+    })
+  })
+
+  describe('Swaps (Fungible)', async () => {
+    let _order
+
+    before('Alice creates an order for Bob (200 AST for 50 DAI)', async () => {
+      _order = await orders.getOrder({
+        signer: {
+          wallet: aliceAddress,
+          token: tokenAST.address,
+          amount: 200,
+        },
+        sender: {
+          wallet: bobAddress,
+          token: tokenDAI.address,
+          amount: 50,
+        },
+      })
+
+      _order.signature = await signatures.getWeb3Signature(
+        _order,
+        aliceAddress,
+        swapAddress,
+        GANACHE_PROVIDER
+      )
+    })
+
+    it('Checks that Bob can swap with Alice (200 AST for 50 DAI)', async () => {
+      emitted(await swap(_order, { from: bobAddress }), 'Swap')
+    })
+
+    it('Checks balances...', async () => {
+      ok(
+        await balances(aliceAddress, [
+          [tokenAST, 800],
+          [tokenDAI, 50],
+        ]),
+        'Alice balances are incorrect'
+      )
+      ok(
+        await balances(bobAddress, [
+          [tokenAST, 200],
+          [tokenDAI, 950],
+        ]),
+        'Bob balances are incorrect'
+      )
+    })
+
+    it('Checks balances...', async () => {
+      ok(
+        await balances(aliceAddress, [
+          [tokenAST, 800],
+          [tokenDAI, 50],
+        ]),
+        'Alice balances are incorrect'
+      )
+      ok(
+        await balances(bobAddress, [
+          [tokenAST, 200],
+          [tokenDAI, 950],
+        ]),
+        'Bob balances are incorrect'
+      )
+    })
+
+    it('Checks that Alice cannot trade more than approved (200 AST)', async () => {
+      const order = await orders.getOrder({
+        signer: {
+          wallet: aliceAddress,
+          token: tokenAST.address,
+          amount: 200,
+        },
+      })
+
+      await reverted(swap(order, { from: bobAddress }))
+    })
+
+    it('Checks that Bob can not trade more than he holds', async () => {
+      const order = await orders.getOrder({
+        signer: {
+          wallet: bobAddress,
+          token: tokenDAI.address,
+          amount: 1000,
+        },
+        sender: {
+          wallet: aliceAddress,
+        },
+      })
+
+      order.signature = await signatures.getWeb3Signature(
+        order,
+        bobAddress,
+        swapAddress,
+        GANACHE_PROVIDER
+      )
+
+      await reverted(swap(order, { from: aliceAddress }), 'TRANSFER_FAILED')
+    })
+
+    it('Checks remaining balances and approvals', async () => {
+      ok(
+        await balances(aliceAddress, [
+          [tokenAST, 800],
+          [tokenDAI, 50],
+        ]),
+        'Alice balances are incorrect'
+      )
+      ok(
+        await balances(bobAddress, [
+          [tokenAST, 200],
+          [tokenDAI, 950],
+        ]),
+        'Bob balances are incorrect'
+      )
+      // Alice and Bob swapped 200 AST for 50 DAI above, thereforeL
+      // Alice's 200 AST approval is now all gone
```

Comment:
Alice approved 250 not 200, as shown by the fact that alice's allowance in the lines below this is 50 not 0